packages/database/src/changeHandlers/ChangeHandler.js

```diff
@@ -0,0 +1,115 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import assert from 'assert';
+
+export class ChangeHandler {
+  /**
+   * A map of change pushers by record type. Each pusher can return a list of changes that will be
+   * queued and scheduled for handling at a later stage, when changes stop coming through.
+   *
+   * If no pusher is defined for the record type, then the `changeDetails` generated by the db
+   * will be simply added to the change queue
+   *
+   * @protected
+   * @type {Record<string, Function>}
+   */
+  changePushers = {};
```

Comment:
Would suggest here that the functions aren't actually doing the pushing, they're doing more of a mapping from one format of change details to another

```suggestion
  /**
   * A map of change translators by record type. Each translator can alter the change details that will be
   * queued and scheduled for handling at a later stage, when changes stop coming through.
   *
   * If no translator is defined for the record type, then the `changeDetails` generated by the db
   * will be simply added to the change queue
   *
   * @protected
   * @type {Record<string, Function>}
   */
  changeTranslators = {};
```

Reply:
I really like that suggestion, since the actual DB changes are not necessarily the ones pushed into the change queue. For example, if we'd use this base class in `EntityHierarchyCacher`, the items pushed would be hierarchy subtree rebuild jobs.

I tried to emphasise this by contrasting `changeDetails` with `changeQueue`, but "translators" makes it even clearer!